### PR TITLE
Action Policies Fallback

### PR DESCRIPTION
**Pull request by filmakarov**
Implementing action policies fallback approach by @hazim-j from [here](https://github.com/stackup-wallet/smartsessions/pull/1) with small edits

### CODE REVIEW

**Comment by zeroknots** [comment]
we should double check codesizelimit with these changes

**Comment by filmakarov** [comment]
> we should double check codesizelimit with these changes

great catch. it's over the limit now

**Comment by filmakarov** [comment]
I think if we agree on the approach in general we can merge it, and then refactor for smaller code size in a separate branch

File: contracts/lib/PolicyLib.sol [hunk 1]
```diff
@@ -55,33 +55,83 @@ library PolicyLib {
...
             // Intersect the validation data from this policy with the accumulated result
-            vd = vd.intersect(_vd);
+            vd = vd.intersect(policies[i]._checkPolicy(permissionId, callOnIPolicy));
         }
+        
+        // Make sure policies can't alter the control flow
+        if (vd == RETRY_WITH_FALLBACK) revert();
     }
 
+    function _checkPolicy(
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:113]: I'd either remove the _ here, or make it private.

File: contracts/lib/PolicyLib.sol [hunk 2]
```diff
@@ -55,33 +55,83 @@ library PolicyLib {
...
 
+        // Iterate over all policies and intersect the validation data
+        for (uint256 i; i < length; i++) {
             // Intersect the validation data from this policy with the accumulated result
-            vd = vd.intersect(_vd);
+            vd = vd.intersect(policies[i]._checkPolicy(permissionId, callOnIPolicy));
         }
+        
+        // Make sure policies can't alter the control flow
+        if (vd == RETRY_WITH_FALLBACK) revert();
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:110]: we should add an error here

File: contracts/lib/PolicyLib.sol [hunk 3]
```diff
@@ -55,33 +55,83 @@ library PolicyLib {
...
+        PermissionId permissionId,
+        bytes memory callOnIPolicy
+    )
+        internal
+        returns (ValidationData _vd)
+    {
+        
+        // Call the policy contract with the provided calldata
+        (bool success, bytes memory returnDataFromPolicy) = policy.excessivelySafeCall({_gas: gasleft(), _value: 0, _maxCopy: 32, _calldata: callOnIPolicy });
+        if (!success) revert();
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:124]: consider adding error message here

File: contracts/lib/PolicyLib.sol [hunk 4]
```diff
@@ -55,33 +55,83 @@ library PolicyLib {
...
+        uint256 minPolicies
+    )
+        internal
+        returns (ValidationData vd)
+    {
+        // Get the list of policies for the given permissionId and account
+        address[] memory policies = $self.policyList[permissionId].values({ account: userOp.sender });
+        uint256 length = policies.length;
+
+        // Ensure the minimum number of policies is met
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:98]: I think this needs clearer docs

File: contracts/lib/PolicyLib.sol [hunk 5]
```diff
@@ -55,33 +55,83 @@ library PolicyLib {
...
+        }
 
+        // Iterate over all policies and intersect the validation data
+        for (uint256 i; i < length; i++) {
             // Intersect the validation data from this policy with the accumulated result
-            vd = vd.intersect(_vd);
+            vd = vd.intersect(policies[i]._checkPolicy(permissionId, callOnIPolicy));
         }
+        
+        // Make sure policies can't alter the control flow
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:109]:
```diff
- // Make sure policies can't alter the control flow
+ // Prevent a malfunctioning policy, to return the magic value RETRY_WITH_FALLBACK and change control flow
```

File: contracts/lib/PolicyLib.sol [hunk 6]
```diff
@@ -128,14 +178,25 @@ library PolicyLib {
...
-        vd = $policies[actionId].check({
+        vd = $policies[actionId].tryCheck({
             userOp: userOp,
             permissionId: permissionId,
             callOnIPolicy: abi.encodeCall(
                 IActionPolicy.checkAction, (permissionId.toConfigId(actionId), userOp.sender, target, value, callData)
             ),
             minPolicies: minPolicies
         });
+        if (vd == RETRY_WITH_FALLBACK) {
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:189]: consider adding some inline docs here:

- why is this done

File: contracts/lib/PolicyLib.sol [hunk 7]
```diff
@@ -55,33 +55,83 @@ library PolicyLib {
...
+     * Same as check but will not revert if minimum number of policies is not met.
+     * This allows a second check with the FALLBACK_ACTIONID.
+     *
+     * @param $self The Policy storage struct containing the list of policies.
+     * @param userOp The PackedUserOperation to be validated.
+     * @param permissionId The identifier for the permission being checked.
+     * @param callOnIPolicy The encoded function call data to be executed on each policy contract.
+     * @param minPolicies The minimum number of policies that must be present and checked.
+     *
+     * @return vd The intersected ValidationData result from all policy checks.
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:82]:
```diff
-      * @return vd The intersected ValidationData result from all policy checks.
+      * @return vd This value could either be:
+      *             - RETRY_WITH_FALLBACK if no adequate policies were set for this permissionId. Signaling the caller, that the Policy fallback procedure SHOULD be used
+      *             - Intersected Validation data of policies. 

```

File: contracts/lib/PolicyLib.sol [hunk 8]
```diff
@@ -128,14 +178,25 @@ library PolicyLib {
...
         // Check the relevant action policy
-        vd = $policies[actionId].check({
+        vd = $policies[actionId].tryCheck({
             userOp: userOp,
             permissionId: permissionId,
             callOnIPolicy: abi.encodeCall(
                 IActionPolicy.checkAction, (permissionId.toConfigId(actionId), userOp.sender, target, value, callData)
             ),
             minPolicies: minPolicies
         });
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:188]:
```diff
+ // if tryCheck returned RETRY_WITH_FALLBACK, the FALLBACK_ACTIONID should be checked.
```

File: contracts/lib/PolicyLib.sol [hunk 9]
```diff
@@ -128,14 +178,25 @@ library PolicyLib {
...
+        if (vd == RETRY_WITH_FALLBACK) {
+            vd = $policies[FALLBACK_ACTIONID].check({
```
Comment (zeroknots) [inline contracts/lib/PolicyLib.sol:190]:
```diff
+ // This will only pass, If the no FALLBACK_ACTIONID was set for this PermissionID, this will revert
```

**Comment by zeroknots** [comment]
appreciate the contribution @hazim-j this is a cool idea

**Comment by hazim-j** [comment]
> appreciate the contribution @hazim-j this is a cool idea

No worries! Excited to see this implemented into the main repo 🙌

File: contracts/lib/PolicyLib.sol (hunk 9 above)
Comment (filmakarov) [reply contracts/lib/PolicyLib.sol:190]: I think I know, what you meant here :)

**Review by zeroknots — APPROVED** [review]

**Pull request merged**